### PR TITLE
external_storage: add GCS support

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Signed-off-by: Yi Wu <yiwu@pingcap.com>

###  What have you changed?
Adding GCS support to external_storage crate. Due to lack of official GCS rust lib, I'm using the tame-gcs lib, which require to use together with a http client lib. The code is mostly following examples in the tame-gcs repository. 

###  What is the type of the changes?
New feature

###  How is the PR tested?
Manual test with the scli tool.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
no

###  Does this PR affect `tidb-ansible`?
no

### CODE REVIEW

File: components/external_storage/Cargo.toml [hunk 1]
```diff
@@ -6,14 +6,20 @@ publish = false
 
 [dependencies]
 rand = "0.7"
+serde = "*"
```
Comment (tennix) [inline components/external_storage/Cargo.toml]: Please use a specific version for the dependencies.

**Comment by yiwu-arbug** [comment]
@kennytm looks like we are depending on https://github.com/EmbarkStudios/tame-gcs/issues/17. It shouldn't be difficult to make tame-gcs's `Multipart` implement the `AsyncRead` trait though. Or do you have better way to get around it?

**Comment by kennytm** [comment]
@yiwu-arbug you can turn a `futures_io::AsyncRead` to a `std::io::Read` using `.compat()`. This is because `tokio01::AsyncRead` inherits `std::io::Read`.

**Comment by DanielZhangQD** [comment]
@kennytm @tennix @overvenus This PR is ready for review now, PTAL, thanks!

File: components/external_storage/src/gcs.rs [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+            return Err(Error::new(ErrorKind::InvalidInput, "missing credentials"));
+        }
+        let svc_info = ServiceAccountInfo::deserialize(&config.credentials_blob).map_err(|e| {
+            Error::new(
+                ErrorKind::InvalidInput,
+                format!(
+                    "invalid credentials_blob {}: {}",
+                    config.credentials_blob, e
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply components/external_storage/src/gcs.rs:53]: good catch!

File: components/external_storage/src/gcs.rs [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    }
+
+    fn maybe_prefix_key(&self, key: &str) -> String {
+        if !self.config.prefix.is_empty() {
+            return format!("{}/{}", self.config.prefix, key);
+        }
+        key.to_owned()
+    }
+
+    fn convert_request<R: 'static>(&self, req: http::Request<R>) -> Result<reqwest::Request>
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply components/external_storage/src/gcs.rs:81]: `req.into()` doesn't support converting from a `Stream` though.

**Pull request merged**